### PR TITLE
Remove the API version in the validator component

### PR DESCRIPTION
**Pull request by fabpot**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13458 |
| License | MIT |
| Doc PR | n/a |

### CODE REVIEW

File: src/Symfony/Component/Validator/Validator/RecursiveValidator.php [hunk 1]
```diff
@@ -112,6 +115,22 @@ public function hasMetadataFor($object)
      */
     public function validate($value, $constraints = null, $groups = null)
     {
+        // The following code must be removed in 3.0
+        $numArgs = func_num_args();
+        if ($numArgs > 3) {
+            trigger_error('The '.__METHOD__.' method is deprecated in version 2.5 and will be removed in version 3.0. Use the Symfony\Component\Validator\Validator\ValidatorInterface::validate method instead.', E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/Validator/Validator/RecursiveValidator.php]: the method itself is not deprecated. Only the signature with more arguments

File: src/Symfony/Component/Validator/Validator/RecursiveValidator.php [hunk 2]
```diff
@@ -137,4 +156,22 @@ public function validatePropertyValue($objectOrClass, $propertyName, $value, $gr
             ->validatePropertyValue($objectOrClass, $propertyName, $value, $groups)
             ->getViolations();
     }
+
+    /**
+     * @internal
```
Comment (stof) [inline src/Symfony/Component/Validator/Validator/RecursiveValidator.php]: no need to mark as internal. It is private already

File: src/Symfony/Component/Validator/Validator/RecursiveValidator.php [hunk 3]
```diff
@@ -137,4 +156,22 @@ public function validatePropertyValue($objectOrClass, $propertyName, $value, $gr
...
+     * @deprecated since version 2.5, to be removed in 3.0.
```
Comment (stof) [inline src/Symfony/Component/Validator/Validator/RecursiveValidator.php]: no need to mark as deprecated. It is a private utility method

**Comment by stof** [comment]
The RecursiveValidator needs to implement the legacy interface too (which is currently implemented only in the LegacyValidator child class). Otherwise we won't provide the BC layer

**Comment by stof** [comment]
The code of the LegacyExecutionContextFactory should also be moved to the code being used. Otherwise we would not be providing a BC layer anymore

**Comment by saro0h** [comment]
@stof the ``LegacyExecutionContextFactory`` is only used in ``src/Symfony/Component/Validator/Tests/Validator/LegacyValidatorLegacyApiTest.php`` and ``src/Symfony/Component/Validator/Tests/Validator/LegacyValidator2Dot5ApiTest.php``. So what do you mean ? Remove the code from here https://github.com/symfony/symfony/blob/2.7/src/Symfony/Component/Validator/Context/LegacyExecutionContextFactory.php#L64-L73 ?

**Comment by stof** [comment]
@saro0h no, it was also used in the 2.5-BC API level (which is removed by this PR). And the goal of this PR should be to merge the 2.5-BC level into the 2.5 one, not to remove it. The 2.5 level alone cannot be used in 2.7 by default, because it is a BC break (the 2.5 API level actually is the 3.0 level, i.e. the new API without the BC layers). The goal of bumping the PHP requirement to 5.3.9 was precisely to be able to use the 2.5-bc layer all the time instead of forcing to choose between 2.4 and 2.5 on old PHP versions

**Comment by fabpot** [comment]
@stof Can you, by any chance, finish this one?

**Comment by stof** [comment]
If it can wait until this weekend, yes. I can work on it.

**Comment by fabpot** [comment]
@stof I think it can wait a few days, sure. Thanks.

**Comment by saro0h** [comment]
Any updates in this one?

**Comment by stof** [comment]
I haven't worked on it yet (it is still in my TODO list though)

**Comment by fabpot** [comment]
@stof Do you think you will have time soon?

**Comment by stof** [comment]
Closing in favor of https://github.com/symfony/symfony/pull/14016

**Pull request closed without merging**